File: frontend/src/types/networkGraph.ts
export interface NetworkGraphNode {
  id: string;
  label: string;
  ip: string;
  node_type: string;
  is_threat: boolean;
  threat_count: number;
  flow_count: number;
  total_bytes: number;
  x: number;
  y: number;
}

export interface NetworkGraphEdge {
  id: string;
  source: string;
  target: string;
  weight: number;
  protocol: string;
  is_threat: boolean;
  total_bytes: number;
}

export interface NetworkGraphMetadata {
  directed: boolean;
  multigraph: boolean;
  layout_algorithm: string;
  node_count: number;
  edge_count: number;
  period_hours: number;
  generated_at: string;
}

export interface NetworkGraphData {
  nodes: NetworkGraphNode[];
  links: NetworkGraphEdge[];
  metadata: NetworkGraphMetadata;
}

export interface NetworkGraphResponse {
  success: boolean;
  message: string;
  data: NetworkGraphData;
  networkx_format: {
    directed?: boolean;
    multigraph?: boolean;
    graph?: Record<string, unknown>;
    nodes?: Array<Record<string, unknown>>;
    links?: Array<Record<string, unknown>>;
  };
}

File: backend/schemas/network_graph.py
from datetime import datetime, timezone
from typing import Any, Optional

from pydantic import BaseModel, Field


class NetworkGraphNode(BaseModel):
  id: str
  label: str
  ip: str
  node_type: str = "external"
  is_threat: bool = False
  threat_count: int = 0
  flow_count: int = 0
  total_bytes: int = 0
  x: float = 0.0
  y: float = 0.0


class NetworkGraphEdge(BaseModel):
  id: str
  source: str
  target: str
  weight: int = 1
  protocol: str = "unknown"
  is_threat: bool = False
  total_bytes: int = 0


class NetworkGraphMetadata(BaseModel):
  directed: bool = True
  multigraph: bool = False
  layout_algorithm: str = "spring"
  node_count: int
  edge_count: int
  period_hours: int
  generated_at: datetime


class NetworkGraphData(BaseModel):
  nodes: list[NetworkGraphNode]
  links: list[NetworkGraphEdge]
  metadata: NetworkGraphMetadata


class NetworkGraphResponse(BaseModel):
  success: bool = True
  message: str
  data: NetworkGraphData
  networkx_format: dict[str, Any] = Field(
    default_factory=dict,
    description="Raw NetworkX node-link export for interoperability",
  )
